util/src/openssl_utils.cpp

```diff
@@ -0,0 +1,258 @@
+//
+// Copyright (c) 2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the LICENSE
+// file.
+
+#include "openssl_utils.hpp"
+#include "Logger.hpp"
+#include "assertUtils.hpp"
+#include "hex_tools.h"
+#include "openssl_crypto.hpp"
+#include "crypto/eddsa/EdDSA.h"
+
+#include <regex>
+#include <utility>
+
+namespace concord::crypto::openssl {
+
+using concord::util::openssl_utils::UniquePKEY;
+using concord::util::openssl_utils::UniqueOpenSSLPKEYContext;
+using concord::util::openssl_utils::UniqueOpenSSLX509;
+using concord::util::openssl_utils::UniqueOpenSSLBIO;
+
+string CertificateUtils::generateSelfSignedCert(const string& origin_cert_path,
+                                                const string& public_key,
+                                                const string& signing_key) {
+  auto deleter = [](FILE* fp) {
+    if (fp) fclose(fp);
+  };
+  unique_ptr<FILE, decltype(deleter)> fp(fopen(origin_cert_path.c_str(), "r"), deleter);
+  if (!fp) {
+    LOG_ERROR(OPENSSL_LOG, "Certificate file not found, path: " << origin_cert_path);
+    return string();
+  }
+
+  UniqueOpenSSLX509 cert(PEM_read_X509(fp.get(), NULL, NULL, NULL));
+  if (!cert) {
+    LOG_ERROR(OPENSSL_LOG, "Cannot parse certificate, path: " << origin_cert_path);
+    return string();
+  }
+
+  UniquePKEY priv_key(EVP_PKEY_new());
+  UniqueOpenSSLBIO priv_bio(BIO_new(BIO_s_mem()));
+
+  if (BIO_write(priv_bio.get(), static_cast<const char*>(signing_key.c_str()), signing_key.size()) <= 0) {
+    LOG_ERROR(OPENSSL_LOG, "Unable to create private key object");
+    return string();
+  }
+
+  if (!PEM_read_bio_PrivateKey(priv_bio.get(), reinterpret_cast<EVP_PKEY**>(&priv_key), NULL, NULL)) {
+    LOG_ERROR(OPENSSL_LOG, "Unable to create private key object");
+    return string();
+  }
+  UniquePKEY pub_key(EVP_PKEY_new());
+  UniqueOpenSSLBIO pub_bio(BIO_new(BIO_s_mem()));
+
+  if (BIO_write(pub_bio.get(), static_cast<const char*>(public_key.c_str()), public_key.size()) <= 0) {
+    LOG_ERROR(OPENSSL_LOG, "Unable to create public key object");
+    return string();
+  }
+  if (!PEM_read_bio_PUBKEY(pub_bio.get(), reinterpret_cast<EVP_PKEY**>(&pub_key), NULL, NULL)) {
+    LOG_ERROR(OPENSSL_LOG, "Unable to create public key object");
+    return string();
+  }
+
+  X509_set_pubkey(cert.get(), pub_key.get());
+  X509_sign(cert.get(), priv_key.get(), EVP_sha256());
+
+  UniqueOpenSSLBIO outbio(BIO_new(BIO_s_mem()));
+  if (!PEM_write_bio_X509(outbio.get(), cert.get())) {
+    LOG_ERROR(OPENSSL_LOG, "Unable to create certificate object");
+    return string();
+  }
+  string certStr;
```

Comment:
Use a vector of bytes instead since you can specify its size in a single statement on construction.

Reply:
The `generateSelfSignedCert()` returns `string` object. This `string` object is used in multiple places as `string`.
We need to modify in multiple places if the return type is made to `vector<uint8_t>`.